perf/src/recycler.rs

```diff
@@ -73,41 +135,108 @@ impl<T: Default + Reset + Sized> Recycler<T> {
+                    && object_pool.len() > shrink_threshold_count as usize
+                {
+                    object_pool.above_shrink_ratio_count += 1;
+                } else {
+                    object_pool.above_shrink_ratio_count = 0;
+                }
 
-        let mut t = T::default();
-        t.set_recycler(Arc::downgrade(&self.recycler));
-        t
+                if object_pool.above_shrink_ratio_count as usize
+                    >= object_pool.max_above_shrink_ratio_count as usize
+                {
+                    // Do the shrink
+                    let total_alocated_count_shrink_target =
+                        std::cmp::max(object_pool.minimum_object_count, shrink_threshold_count);
+                    let target_num_to_shrink =
+                        object_pool.total_allocated_count - total_alocated_count_shrink_target;
+                    for _ in 0..target_num_to_shrink {
+                        if let Some(mut expired_object) = object_pool.object_pool.pop() {
+                            expired_object.unset_recycler();
+                            // May not be able to shrink exactly `target_num_to_shrink` objects sinc
+                            // in the case of new allocations, `total_allocated_count` is incremented
+                            // before the object is allocated (see `should_allocate_new` logic below).
+                            // This race allows a difference of up to the number of threads allocating
+                            // with this recycler.
+                            object_pool.total_allocated_count -= 1;
+                        }
+                    }
+
+                    object_pool.above_shrink_ratio_count = 0;
+                    datapoint_info!(
+                        "recycler_shrink",
+                        (
+                            "total_alocated_count_shrink_target",
+                            total_alocated_count_shrink_target as i64,
+                            i64
+                        ),
+                        ("target_num_to_shrink", target_num_to_shrink as i64, i64),
+                        (
+                            "total_allocated_count",
+                            object_pool.total_allocated_count as i64,
+                            i64
+                        ),
+                    );
+                }
+            }
+
+            let reused_object = object_pool.object_pool.pop();
+            if reused_object.is_some() {
+                (reused_object, true, false)
+            } else if let Some(limit) = object_pool.limit {
+                let should_allocate_new = object_pool.total_allocated_count < limit;
+                if should_allocate_new {
+                    object_pool.total_allocated_count += 1;
+                }
+                (None, false, should_allocate_new)
+            } else {
+                (None, false, true)
+            }
+        };
```

Comment:
@ryoqun this elastic shrinking seems to be working well for my test case where I spam the node, I can see the total number of objects rising then falling over time after the spam is over. 

What do you think?

Reply:
@carllin cool. this looks promising by superseding my attempt #15139. After all #15139 wasn't prepared for spamming, it's half way in that it only addresses spontaneous spikes, not spamming... ;)